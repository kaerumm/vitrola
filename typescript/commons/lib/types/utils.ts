export type Nullable<T> = T | null

export type Expect<T extends true> = T
export type ExpectFalse<T extends false> = T

export type Equal<X, Y> =
    (<T>() => T extends X ? 1 : 2) extends <T>() => T extends Y ? 1 : 2
        ? true
        : false
export type NotEqual<X, Y> = true extends Equal<X, Y> ? false : true

// https://stackoverflow.com/questions/49927523/disallow-call-with-any/49928360#49928360
export type IsAny<T> = 0 extends 1 & T ? true : false
export type NotAny<T> = true extends IsAny<T> ? false : true

export type Debug<T> = { [K in keyof T]: T[K] }
export type MergeInsertions<T> = T extends object
    ? { [K in keyof T]: MergeInsertions<T[K]> }
    : T

export type Alike<X, Y> = Equal<MergeInsertions<X>, MergeInsertions<Y>>
export type ExpectExtends<Value, Expected> = Expected extends Value
    ? true
    : false
export type ExpectValidArgs<
    Func extends (...args: any[]) => any,
    Args extends any[],
> = Args extends Parameters<Func> ? true : false
export type UnionToIntersection<U> = (
    U extends any ? (k: U) => void : never
) extends (k: infer I) => void
    ? I
    : never

export type Split<
    Value extends string,
    Separator extends string = '.',
> = Value extends `${infer Left}${Separator}${infer Right}`
    ? [Left, ...Split<Right, Separator>]
    : [Value]

type SplitTest = [
    Expect<Equal<Split<'a.b.c', '.'>, ['a', 'b', 'c']>>,
    Expect<Equal<Split<'a,b,c', ','>, ['a', 'b', 'c']>>,
]

export type Reverse<Value extends any[]> = Value extends [
    ...infer Rest,
    infer Last,
]
    ? [Last, ...Reverse<Rest>]
    : Value

type ReverseTest = [Expect<Equal<Reverse<[1, 2, 3]>, [3, 2, 1]>>]

/**
 *
 * Erases the type if it is an object and it is a function, a symbol or an array
 * This can be used when you want only map/dictionary objects
 */
export type MapObject<O> = O extends object
    ? O extends Function
        ? never
        : O extends Symbol
          ? never
          : O extends Array<any>
            ? never
            : O
    : never

export type ArgumentsOf<T> = T extends (...args: infer Arguments) => any
    ? Arguments
    : undefined

type ArgumentsOfTest = [
    Expect<
        Equal<
            ArgumentsOf<
                (
                    a: number,
                    b: boolean,
                    c: 'string',
                    d: 0,
                    e: [0, boolean, { a: 0; b: boolean }],
                    f: { a: 0; b: boolean }
                ) => void
            >,
            [
                number,
                boolean,
                'string',
                0,
                [0, boolean, { a: 0; b: boolean }],
                { a: 0; b: boolean },
            ]
        >
    >,
    Expect<Equal<ArgumentsOf<string>, undefined>>,
]

/**
 *
 * Matches U or some narrower value of U
 *
 * For instance if U is number, matches number or any specific number like 0, 1, etc...
 */
export type KeysMatchingSubtype<T extends object, U> = {
    [Key in keyof T]-?: T[Key] extends U ? Key : never
}[keyof T]

type KeysMatchingSubtypeTest = [
    Expect<
        Equal<
            KeysMatchingSubtype<
                { narrow: 0; exact: number; wide: number | string },
                number
            >,
            'narrow' | 'exact'
        >
    >,
]

/**
 *
 * Matches U or some narrower value of U
 *
 * For instance if U is number, matches number or any specific number like 0, 1, etc...
 */
export type KeysMatchingExactType<T extends object, U> = {
    [Key in keyof T]-?: T[Key] extends U
        ? [U] extends [T[Key]]
            ? Key
            : never
        : never
}[keyof T]

type KeysMatchingExactTypeTest = [
    Expect<
        Equal<
            KeysMatchingExactType<
                { narrow: 0; exact: number; wide: number | string },
                number
            >,
            'exact'
        >
    >,
]

/**
 *
 * Matches U or some union of U
 *
 */
export type KeysMatchingSupertype<T extends object, U> = {
    [Key in keyof T]-?: [U] extends [T[Key]] ? Key : never
}[keyof T]

type KeysMatchingSupertypeTest = [
    Expect<
        Equal<
            KeysMatchingSupertype<
                { narrow: 0; exact: number; wide: number | string },
                number
            >,
            'exact' | 'wide'
        >
    >,
]

// not used
type ExtendsStrict<T, U> = T extends U ? ([U] extends [T] ? T : never) : never

export interface Constructor<T, Args extends any[] = []> {
    new (...args: Args): T
}

export type Join<
    Array extends string[],
    Separator extends string,
> = Array extends [infer First extends string, ...infer Rest extends string[]]
    ? Rest['length'] extends 0
        ? `${First}`
        : `${First}${Separator}${Join<Rest, Separator>}`
    : never

type JoinTest = [
    Expect<Equal<Join<['a', 'b'], '.'>, 'a.b'>>,
    Expect<Equal<Join<['a'], '.'>, 'a'>>,
]
