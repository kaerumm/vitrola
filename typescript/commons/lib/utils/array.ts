import type { Option } from './option'

/**
 * @param array Array must be ordered from the lowest to the highest
 */
export function binarySearch(
    array: readonly number[],
    target: number
): Option<number> {
    if (array.length === 0) {
        return null
    }
    let low = 0
    let high = array.length - 1
    let mid: number
    do {
        mid = Math.floor((high + low) / 2)
        if (array[mid] === target) {
            return mid
        }
        if (target < array[mid]) {
            high = mid - 1
        } else {
            low = mid + 1
        }
    } while (mid !== low && mid !== high)
    return null
}

export function* enumerated<T>(array: T[]): Generator<[T, number]> {
    for (let index = 0; index < array.length; index++) {
        yield [array[index], index]
    }
}
