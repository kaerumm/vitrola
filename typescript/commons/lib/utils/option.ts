type Some<T> = T extends void ? undefined : T
type None = null
export type Option<T> = Some<T> | None

export class Options {
    private constructor() {}

    static some<T>(value: Some<T>): Option<T> {
        return value as Option<T>
    }

    static none<T>(): Option<T> {
        return null as Option<T>
    }

    static fromTruthy<T>(value?: T): Option<T> {
        if (!value) {
            return null
        }
        return value as Some<T>
    }

    static isSome<T>(option: Option<T>): option is Some<T> {
        return option !== null
    }

    static isNone<T>(option: Option<T>): option is None {
        return option === null
    }

    static map<T, U>(option: Option<T>, mapper: (value: T) => U): Option<U> {
        if (this.isNone(option)) {
            return null
        }
        return mapper(option as T) as Some<U>
    }
}
