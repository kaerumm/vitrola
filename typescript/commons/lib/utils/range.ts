export type Range = [from: number, to: number]

/**
 * @param from Inclusive
 * @param to Not inclusive
 *
 *  [from, to)
 */
export function range(
    from: number,
    to: number,
    step: number = 1,
    inclusive = false
) {
    const direction = (to - from) / Math.abs(to - from)
    if (inclusive) {
        to = to + direction
    }
    if (direction > 0) {
        return positiveDirection(from, to, direction * step)
    }
    return negativeDirection(from, to, direction * step)
}

function* negativeDirection(from: number, to: number, step: number) {
    for (let i = from; i > to; i += step) {
        yield i
    }
}

function* positiveDirection(from: number, to: number, step: number) {
    for (let i = from; i < to; i += step) {
        yield i
    }
}
