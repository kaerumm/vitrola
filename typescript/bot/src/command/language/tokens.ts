import { Option } from 'commons/lib/utils/option'

// Literals
export interface StringToken {
    kind: 'string'
    value: string
}

// One or two character tokens
export interface LeftParen {
    kind: 'left_paren'
}
export interface RightParen {
    kind: 'right_paren'
}
export interface Minus {
    kind: 'minus'
}
export interface Plus {
    kind: 'plus'
}
export interface Semicolon {
    kind: 'semicolon'
}
export interface ForwardSlash {
    kind: 'forward_slash'
}
export interface Star {
    kind: 'star'
}
export interface Bang {
    kind: 'bang'
}
export interface BangEqual {
    kind: 'bang_equal'
}
export interface Equal {
    kind: 'equal'
}
export interface EqualEqual {
    kind: 'equal_equal'
}
export interface Greater {
    kind: 'greater'
}
export interface GreaterEqual {
    kind: 'greater_equal'
}
export interface Less {
    kind: 'less'
}
export interface LessEqual {
    kind: 'less_equal'
}
export interface Pipe {
    kind: 'pipe'
}
export interface And {
    kind: 'and'
}
export interface Or {
    kind: 'or'
}

export type Token =
    | LeftParen
    | RightParen
    | StringToken
    | Minus
    | Plus
    | Semicolon
    | ForwardSlash
    | Star
    | Bang
    | BangEqual
    | Equal
    | EqualEqual
    | Greater
    | GreaterEqual
    | Less
    | LessEqual
    | Pipe
    | And
    | Or

export interface Span<T extends Token = Token> {
    token: T
    position: number
}

export interface SyntaxError {
    kind: 'unfinished_string'
    position: number
}

export function isBinaryOperator(
    span: Option<Span<Token>>
): span is Span<And | Or> {
    return !!span && ['and', 'or'].includes(span.token.kind)
}

export function isStringSpan(span: Span<Token>): span is Span<StringToken> {
    return span.token.kind === 'string'
}

export function isGroupingSpan(span: Span<Token>): span is Span<LeftParen> {
    return span.token.kind === 'left_paren'
}
