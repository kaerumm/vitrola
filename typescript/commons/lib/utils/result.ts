import { Options, type Option } from './option'

export type ValueResult<T> = T extends void ? undefined : T
export interface ErrorResult<E> {
    error: E
}
export type Result<T, E = Error> = ValueResult<T> | ErrorResult<E>

export type NotError<T> = T extends ErrorResult<any> ? never : T

export class Results {
    private constructor() {}

    static wrapError<T>(fn: () => T): Result<T, unknown> {
        try {
            return fn() as ValueResult<T>
        } catch (error: unknown) {
            return { error }
        }
    }

    static async wrapErrorAsync<T>(
        fn: () => Promise<T>
    ): Promise<Result<T, unknown>> {
        try {
            return (await fn()) as ValueResult<T>
        } catch (error: unknown) {
            return { error }
        }
    }

    static error<E>(error: E): ErrorResult<E> {
        return { error }
    }

    static ok<T>(value: ValueResult<T>): ValueResult<T> {
        return value
    }

    static isOk<T>(result: Result<T, any>): result is ValueResult<T> {
        return !this.isErr(result)
    }

    static isErr<E>(result: Result<any, E>): result is ErrorResult<E> {
        return !!result && typeof result === 'object' && 'error' in result
    }

    static map<T, E, U>(
        result: Result<T, E>,
        mapper: (result: T) => U
    ): Result<U, E> {
        if (this.isOk(result)) {
            return mapper(result as T) as Result<U, E>
        }
        return result as ErrorResult<E>
    }

    static orElse<T>(result: Result<T, any>, or: T): T {
        if (this.isOk(result)) {
            return result as T
        }
        return or
    }

    static optionErr<E>(result: Result<any, E>) {
        return this.map(result, (result) => {
            if (this.isErr(result)) {
                return Options.some(result.error)
            }
            return Options.none()
        })
    }

    static optionOk<T>(result: Result<T, any>): Option<T> {
        if (this.isOk(result)) {
            return Options.some(result)
        }
        return Options.none()
    }
}
