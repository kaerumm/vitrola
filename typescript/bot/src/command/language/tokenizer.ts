import { Cursor } from 'commons/lib/data-structures/cursor'
import { Results, type Result } from 'commons/lib/utils/result'
import type { Option } from 'commons/lib/utils/option'
import {
    CharacterCodes,
    isSpecialCharacter,
    isUnicodeLineBreak,
    isUnicodeWhitespace,
} from './character'
import type { Span, Token, SyntaxError } from './tokens'

export class Tokenizer {
    static tokenize(source: string): Result<Span<Token>[], SyntaxError[]> {
        const cursor = new Cursor<string>(source)
        const spans: Span<Token>[] = []
        const errors: SyntaxError[] = []
        let character: Option<string>
        while ((character = cursor.next())) {
            let token: Option<Result<Token, SyntaxError>> = null
            let position = cursor.position - 1
            switch (true) {
                case character === '(':
                    token = { kind: 'left_paren' }
                    break
                case character === ')':
                    token = { kind: 'right_paren' }
                    break
                case character === '-':
                    token = { kind: 'minus' }
                    break
                case character === '+':
                    token = { kind: 'plus' }
                    break
                case character === ';':
                    token = { kind: 'semicolon' }
                    break
                case character === '/':
                    token = { kind: 'forward_slash' }
                    break
                case character === '*':
                    token = { kind: 'star' }
                    break
                case character === '!':
                    if (cursor.peek(0) === '=') {
                        cursor.next()
                        token = { kind: 'bang_equal' }
                        break
                    }
                    token = { kind: 'bang' }
                    break
                case character === '=':
                    if (cursor.peek(0) === '=') {
                        cursor.next()
                        token = { kind: 'equal_equal' }
                        break
                    }
                    token = { kind: 'equal' }
                    break
                case character === '>':
                    if (cursor.peek(0) === '=') {
                        cursor.next()
                        token = { kind: 'greater_equal' }
                        break
                    }
                    token = { kind: 'greater' }
                    break
                case character === '<':
                    if (cursor.peek(0) === '=') {
                        cursor.next()
                        token = { kind: 'less_equal' }
                        break
                    }
                    token = { kind: 'less' }
                    break
                case character === '&':
                    if (cursor.peek(0) === '&') {
                        cursor.next()
                        token = { kind: 'and' }
                        break
                    }
                    // TODO: Background running support(?) does that even make sense for our context?
                    break
                case character === '|':
                    if (cursor.peek(0) === '|') {
                        cursor.next()
                        token = { kind: 'or' }
                        break
                    }
                    token = { kind: 'pipe' }
                    break
                case character === "'" || character === '"':
                    token = Results.map(
                        tokenizeString(character, cursor, source, position),
                        (value) => ({
                            kind: 'string',
                            value,
                        })
                    )
                    break
                case isUnicodeWhitespace(character):
                    // Ignore the input
                    break
                case isUnicodeLineBreak(character):
                    consumeLineTerminator(character as string, cursor)
                    token = {
                        kind: 'semicolon',
                    }
                    break
                case !isSpecialCharacter(character):
                    token = {
                        kind: 'string',
                        value: tokenizeUnquotedString(cursor, source, position),
                    }
                    break
                default:
                    // Ignore the input
                    break
            }
            if (!token) {
                continue
            }
            if (Results.isErr(token)) {
                errors.push(token.error)
                continue
            }
            spans.push({ token, position })
        }
        if (errors.length > 0) {
            return { error: errors }
        }
        return spans
    }
}

export function consumeLineTerminator(
    firstTerminator: string,
    cursor: Cursor<string>
): void {
    if (
        firstTerminator.codePointAt(0)! === CharacterCodes.CarriageReturn &&
        cursor.peek(0)?.codePointAt(0)! === CharacterCodes.LineFeed
    ) {
        cursor.next()
    }
}

export function tokenizeUnquotedString(
    cursor: Cursor<string>,
    source: string,
    firstCharacterPosition: number
): string {
    let lastCharacterPosition = firstCharacterPosition
    let next
    do {
        lastCharacterPosition = cursor.position
        if (isSpecialCharacter(cursor.peek(0))) {
            break
        }
        next = cursor.next()
    } while (next)
    return source.slice(firstCharacterPosition, lastCharacterPosition)
}

export function tokenizeString(
    delimiter: "'" | '"',
    cursor: Cursor<string>,
    source: string,
    delimiterPosition: number,
    ignoreError = false
): Result<string, SyntaxError> {
    let next
    let firstCharacterPosition = delimiterPosition + 1
    let lastCharacterPosition = firstCharacterPosition
    do {
        lastCharacterPosition = cursor.position
        next = cursor.next()
        if (!next) {
            if (ignoreError) {
                break
            }
            return {
                error: {
                    kind: 'unfinished_string',
                    position: delimiterPosition,
                },
            }
        }
    } while (next !== delimiter)
    return source.slice(firstCharacterPosition, lastCharacterPosition)
}
