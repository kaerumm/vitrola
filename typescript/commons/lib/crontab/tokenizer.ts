import { Result, Results } from '../utils/result.ts'
import { Cursor } from '../data-structures/cursor.ts'
import { Option } from '../utils/option.ts'
import {
    isDigit,
    isUnicodeLineBreak,
    isUnicodeWhitespace,
} from '../utils/character.ts'
import { parseInteger } from '../utils/number.ts'
import { isAllowedToStartCronKeyword } from './character.ts'
import { unreachable } from '../utils/error.ts'

const months = [
    'jan',
    'feb',
    'mar',
    'apr',
    'may',
    'jun',
    'jul',
    'aug',
    'sep',
    'oct',
    'nov',
    'dec',
    'JAN',
    'FEB',
    'MAR',
    'APR',
    'MAY',
    'JUN',
    'JUL',
    'AUG',
    'SEP',
    'OCT',
    'NOV',
    'DEC',
] as const
const daysOfWeek = [
    'sun',
    'mon',
    'tue',
    'wed',
    'thu',
    'fri',
    'sat',
    'SUN',
    'MON',
    'TUE',
    'WED',
    'THU',
    'FRI',
    'SAT',
] as const
const special = ['@reboot'] as const

export const keywords = [...months, ...daysOfWeek, ...special] as const
const keywordSet = new Set<string>(keywords)

export interface Star {
    kind: 'star'
}

export interface ForwardSlash {
    kind: 'forward_slash'
}

export interface Number {
    kind: 'number'
    value: number
}

export interface Comma {
    kind: 'comma'
}

export interface Dash {
    kind: 'dash'
}

export interface Keyword {
    kind: 'keyword'
    key: (typeof keywords)[number]
}

export type Token = Star | ForwardSlash | Number | Comma | Dash | Keyword

export interface UnexpectedToken {
    kind: 'unexpected_token'
    character: string
}

export interface InvalidNumber {
    kind: 'invalid_number'
    source: string
}

export interface UnknownKeyword {
    kind: 'unknown_keyword'
    word: string
}
export type CrontabTokenizationError =
    | UnexpectedToken
    | InvalidNumber
    | UnknownKeyword

export class Tokenizer {
    static tokenize(source: string): Result<Token[], CrontabTokenizationError> {
        const cursor = new Cursor<string>(source)
        const tokens: Token[] = []
        let character: Option<string>
        while ((character = cursor.next())) {
            let token: Option<Result<Token, CrontabTokenizationError>> = null
            let position = cursor.position - 1
            switch (true) {
                case isAllowedToStartCronKeyword(character):
                    token = this.parseCronKeyword(position, cursor, source)
                    break
                case isDigit(character):
                    token = this.parsePositiveInteger(position, cursor, source)
                    break
                case character === '*':
                    token = { kind: 'star' }
                    break
                case character === '/':
                    token = { kind: 'forward_slash' }
                    break
                case character === ',':
                    token = { kind: 'comma' }
                    break
                case character === '-':
                    token = { kind: 'dash' }
                    break
                case isUnicodeWhitespace(character):
                    break
                case isUnicodeLineBreak(character):
                    return tokens
            }
            if (!token) {
                continue
            }
            if (Results.isErr(token)) {
                return token
            }
            tokens.push(token)
        }
        return tokens
    }

    static parseCronKeyword(
        firstCharacterPosition: number,
        cursor: Cursor<string>,
        source: string
    ): Result<Keyword, CrontabTokenizationError> {
        let next: Option<string>
        while ((next = cursor.next())) {
            if (!isAllowedToStartCronKeyword(next)) {
                break
            }
        }
        let str = source.slice(firstCharacterPosition, cursor.position)
        if (!isKeyword(str)) {
            return Results.error({ kind: 'unknown_keyword', word: str })
        }
        return { kind: 'keyword', key: str }
    }

    static parsePositiveInteger(
        firstDigitPosition: number,
        cursor: Cursor<string>,
        source: string
    ): Result<Number, CrontabTokenizationError> {
        let next: Option<string>
        while ((next = cursor.next())) {
            if (
                !next ||
                isUnicodeWhitespace(next) ||
                isUnicodeLineBreak(next)
            ) {
                break
            }
            if (!isDigit(next)) {
                return Results.error({
                    kind: 'unexpected_token',
                    character: next,
                })
            }
        }
        return Results.map(
            Results.mapError(
                parseInteger(
                    source.slice(firstDigitPosition, cursor.position),
                    10
                ),
                () => ({
                    kind: 'invalid_number',
                    source: source.slice(firstDigitPosition, cursor.position),
                })
            ),
            (value) => ({
                kind: 'number',
                value,
            })
        )
    }

    static tokenizeOrUnreachable(source: string): Token[] {
        return Results.mapOrElse(Tokenizer.tokenize(source), () =>
            unreachable(
                'This function should only be called if the tokenization is not guaranteed to fail'
            )
        )
    }
}

function isKeyword(key: string): key is (typeof keywords)[number] {
    return keywordSet.has(key)
}
