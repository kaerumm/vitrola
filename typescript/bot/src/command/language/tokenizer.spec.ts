import { describe, test, expect } from '@jest/globals'
import { Tokenizer } from './tokenizer'
import * as fs from 'fs'
import * as path from 'path'
import {
    CharacterCodes,
    MAX_UNICODE_CODEPOINT,
    lineBreakers,
    reservedCharacters,
    whitespaces,
} from './character'
import { range } from 'commons/lib/utils/range'
import { PseudoRandomNumberGenerator, fuzzyTest } from 'testing/lib/fuzz/fuzz'
import { Results } from 'commons/lib/utils/result'
import { Option } from 'commons/lib/utils/option'
import { Span } from './tokens'

console.log(path.resolve('./test-files/tokenizer_snapshot_file'))
const snapshot_file = fs.readFileSync(
    path.resolve('./test-files/tokenizer_snapshot_file'),
    'utf-8'
)

class TokenizerFuzzer {
    private static defaultOptions = {
        stringLength: 10,
    }

    private options: typeof TokenizerFuzzer.defaultOptions

    constructor(
        options: Option<Partial<typeof TokenizerFuzzer.defaultOptions>>
    ) {
        this.options = {
            ...TokenizerFuzzer.defaultOptions,
            ...options,
        }
    }

    fuzz(prng: PseudoRandomNumberGenerator): string {
        let str = ''
        for (const _ of range(0, this.options.stringLength)) {
            // Unicode characters are allowed to be from the range 0x0 to 0x10ffff
            const codePoint = Math.round(prng() * 0x10ffff)
            str += String.fromCodePoint(codePoint)
        }
        return str
    }
}

describe('Tokenizer', function () {
    test('Snapshot', function () {
        expect(Tokenizer.tokenize(snapshot_file)).toMatchSnapshot()
    })

    describe('Manual', function () {
        test('Symbols', function () {
            const pairs = [
                [`(`, { token: { kind: 'left_paren' } }],
                [`)`, { token: { kind: 'right_paren' } }],
                [`-`, { token: { kind: 'minus' } }],
                [`+`, { token: { kind: 'plus' } }],
                [`;`, { token: { kind: 'semicolon' } }],
                [`/`, { token: { kind: 'forward_slash' } }],
                ['*', { token: { kind: 'star' } }],
                ['!', { token: { kind: 'bang' } }],
                ['=', { token: { kind: 'equal' } }],
                ['>', { token: { kind: 'greater' } }],
                ['<', { token: { kind: 'less' } }],
                ['!=', { token: { kind: 'bang_equal' } }],
                ['==', { token: { kind: 'equal_equal' } }],
                ['>=', { token: { kind: 'greater_equal' } }],
                ['<=', { token: { kind: 'less_equal' } }],
                ['&&', { token: { kind: 'and' } }],
                ['||', { token: { kind: 'or' } }],
                ['|', { token: { kind: 'pipe' } }],
            ] as const
            for (const pair of pairs) {
                expect(Tokenizer.tokenize(pair[0])).toEqual([
                    {
                        ...pair[1],
                        position: 0,
                    },
                ])
            }
        })

        test('Whitespace', function () {
            const whitespace = whitespaces.map((codePoint) =>
                String.fromCodePoint(codePoint)
            )
            for (const space of whitespace) {
                expect(Tokenizer.tokenize(space)).toStrictEqual([])
            }
        })

        test('Line breakers', function () {
            const breakers = lineBreakers.map((codePoint) =>
                String.fromCodePoint(codePoint)
            )
            breakers.push(
                // A CarriageReturn followed by a LineFeed must be tokenized as a single semicolon.
                `${String.fromCodePoint(CharacterCodes.CarriageReturn)}${String.fromCodePoint(CharacterCodes.LineFeed)}`
            )
            for (const breaker of breakers) {
                expect(Tokenizer.tokenize(breaker)).toStrictEqual([
                    {
                        token: { kind: 'semicolon' },
                        position: 0,
                    },
                ])
            }
        })

        test('Quoted string', function () {
            const delimiters = [`\'`, `\"`]
            // Quoted strings are allowed to have any unicode character, except the starting delimiter,
            // inside of it.
            for (const codePoint of range(0, MAX_UNICODE_CODEPOINT, 1, true)) {
                const char = String.fromCodePoint(codePoint)
                for (const delimiter of delimiters) {
                    if (char === delimiter) {
                        continue
                    }
                    expect(
                        Tokenizer.tokenize(`${delimiter}${char}${delimiter}`)
                    ).toEqual([
                        { token: { kind: 'string', value: char }, position: 0 },
                    ])
                }
            }
            // Unfinished strings must error
            expect(Tokenizer.tokenize(`"""`)).toEqual({
                error: [{ kind: 'unfinished_string', position: 2 }],
            })
            expect(Tokenizer.tokenize(`'''`)).toEqual({
                error: [{ kind: 'unfinished_string', position: 2 }],
            })
        })

        test('Unquoted string', function () {
            for (const whitespace of whitespaces) {
                const character = String.fromCodePoint(whitespace)
                // Must terminate unquoted string
                expect(Tokenizer.tokenize(`word${character}`)).toStrictEqual([
                    {
                        token: {
                            kind: 'string',
                            value: 'word',
                        },
                        position: 0,
                    },
                ])
                // Must not be included if it comes before the word
                expect(Tokenizer.tokenize(`${character}word`)).toEqual([
                    {
                        token: {
                            kind: 'string',
                            value: 'word',
                        },
                        position: 1,
                    },
                ])
            }
            for (const specialCharacter of [
                ...lineBreakers,
                ...reservedCharacters,
            ]) {
                const delimiters = [`\'`, `\"`]
                const character = String.fromCodePoint(specialCharacter)
                if (delimiters.includes(character)) {
                    continue
                }
                // Must terminate unquoted string
                expect(
                    (Tokenizer.tokenize(`word${character}`) as Span[])[0]
                ).toStrictEqual({
                    token: {
                        kind: 'string',
                        value: 'word',
                    },
                    position: 0,
                })
                // Must not be included if it comes before the word
                expect(
                    (Tokenizer.tokenize(`${character}word`) as Span[]).find(
                        (s) =>
                            s.token.kind === 'string' &&
                            s.token.value === 'word'
                    )
                ).toBeTruthy()
            }
        })
    })

    test('Fuzzy test for unexpected crashes', function () {
        expect(
            fuzzyTest(
                new TokenizerFuzzer(null),
                (unicodeString) => {
                    Tokenizer.tokenize(unicodeString)
                },
                { iterations: 100_000 }
            )
        ).toEqual(Results.ok(undefined))
    })
})
