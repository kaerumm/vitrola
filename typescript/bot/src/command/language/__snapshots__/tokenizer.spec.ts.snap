// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokenizer Snapshot 1`] = `
[
  {
    "position": 0,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 7,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 8,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 16,
    "token": {
      "kind": "string",
      "value": "strOrLiteral",
    },
  },
  {
    "position": 28,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 29,
    "token": {
      "kind": "left_paren",
    },
  },
  {
    "position": 30,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 38,
    "token": {
      "kind": "string",
      "value": "strOrLiteral",
    },
  },
  {
    "position": 50,
    "token": {
      "kind": "right_paren",
    },
  },
  {
    "position": 51,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 52,
    "token": {
      "kind": "string",
      "value": "string",
    },
  },
  {
    "position": 60,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 61,
    "token": {
      "kind": "string",
      "value": "string",
    },
  },
  {
    "position": 69,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 70,
    "token": {
      "kind": "string",
      "value": "string still string <>!=|&*/+-;&&||==!=>=<="",
    },
  },
  {
    "position": 116,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 117,
    "token": {
      "kind": "string",
      "value": "string still string <>!=|&*/+-;&&||==!=>=<='",
    },
  },
  {
    "position": 163,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 164,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 171,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 172,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 179,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 180,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 188,
    "token": {
      "kind": "string",
      "value": "strOrLiteral",
    },
  },
  {
    "position": 200,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 201,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 209,
    "token": {
      "kind": "string",
      "value": "strOrLiteral",
    },
  },
  {
    "position": 221,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 222,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 230,
    "token": {
      "kind": "and",
    },
  },
  {
    "position": 233,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 240,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 241,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 249,
    "token": {
      "kind": "string",
      "value": "strOrLiteral",
    },
  },
  {
    "position": 262,
    "token": {
      "kind": "and",
    },
  },
  {
    "position": 265,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 273,
    "token": {
      "kind": "string",
      "value": "strOrLiteral",
    },
  },
  {
    "position": 285,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 286,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 294,
    "token": {
      "kind": "or",
    },
  },
  {
    "position": 297,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 304,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 305,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 313,
    "token": {
      "kind": "string",
      "value": "strOrLiteral",
    },
  },
  {
    "position": 326,
    "token": {
      "kind": "or",
    },
  },
  {
    "position": 329,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 337,
    "token": {
      "kind": "string",
      "value": "strOrLiteral",
    },
  },
  {
    "position": 349,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 350,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 358,
    "token": {
      "kind": "pipe",
    },
  },
  {
    "position": 360,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 367,
    "token": {
      "kind": "semicolon",
    },
  },
  {
    "position": 368,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 376,
    "token": {
      "kind": "string",
      "value": "strOrLiteral",
    },
  },
  {
    "position": 389,
    "token": {
      "kind": "pipe",
    },
  },
  {
    "position": 391,
    "token": {
      "kind": "string",
      "value": "command",
    },
  },
  {
    "position": 399,
    "token": {
      "kind": "string",
      "value": "strOrLiteral",
    },
  },
  {
    "position": 411,
    "token": {
      "kind": "semicolon",
    },
  },
]
`;
